File: 240102/뿌요뿌요/puyo-puyo.cpp
#include <iostream>
#include <vector>
#define MAX_N 100
using namespace std;

int n;
int grid[MAX_N][MAX_N];
bool visited[MAX_N][MAX_N];
vector<int> blocks;
int ans;
int cur_block;

bool InRange(int x, int y){
    return 0 <= x && x < n && 0 <= y && y < n;
}

bool CanGo(int x, int y, int k){
    return InRange(x,y) && !visited[x][y] && grid[x][y] == k;
}

void DFS(int x, int y, int k){
    int dirs[4][2] = {{1,0},{0,1},{-1,0},{0,-1}};
    for(int d=0; d<4; d++){
        int nx = x + dirs[d][0], ny = y + dirs[d][1];
        if(CanGo(nx,ny,k)){
            visited[nx][ny] = true;
            cur_block++;
            DFS(nx,ny,k);
        }
    }
}

int main() {
    cin >> n;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cin >> grid[i][j];
        }
    }

    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            if(!visited[i][j]){
                // 만약 방문하지 않았다면
                cur_block = 1;
                visited[i][j] = true;
                DFS(i, j, grid[i][j]);
                blocks.push_back(cur_block);
            }
        }
    }
    int max_block = 0;
    for(int i=0; i<(int) blocks.size(); i++){
        if(blocks[i] >= 4) ans++;
        max_block = max(max_block, blocks[i]);
    }

    cout << ans << ' ' << max_block;
    
    return 0;
}

File: 240127/뿌요뿌요/puyo-puyo.py
n = int(input())
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]
visited = [[False for _ in range(n)] for _ in range(n)]
dirs = ((1,0),(0,1),(-1,0),(0,-1))

bomb_cnt = 0
max_block = 0
cur_block = 0

# 탐색하는 위치가 격자 범위 내에 있는지 여부를 반환
def InRange(x, y):
    return 0 <= x and x < n and 0 <= y and y < n

# 탐색하는 위치로 움직일 수 있는지 여부를 반
def CanGo(x, y, val):
    return InRange(x,y) and not visited[x][y] and grid[x][y] == val

def dfs(x,y,val):
    global cur_block
    # 네 방향 각각에 대하여 DFS 탐색을 하기
    for d in range(4):
        nx,ny = x+dirs[d][0], y+dirs[d][1]
        if CanGo(nx,ny,val):
            # 블럭이 하나 추가됨
            cur_block += 1
            visited[nx][ny] = True
            dfs(nx,ny,val)

# 격자의 각 위치에서 탐색을 시작할 수 있는 경우
# 한 블럭에 대한 DFS 탐색을 수행
for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            # 해당 블럭을 방문할 수 있는 경우 visited 배열을 갱신하고
            # 새로운 블럭을 탐색한다는 의미로 cur_block을 1로 갱신
            cur_block = 1
            visited[i][j] = True
            dfs(i,j,grid[i][j])

            # 한 블럭 묶음에 대한 탐색이 끝난 경우 답을 갱신하기
            if cur_block >= 4:
                bomb_cnt += 1
            
            max_block = max(max_block, cur_block)

print(bomb_cnt, max_block)